### PR TITLE
nvbios: better voltage table parsing

### PR DESCRIPTION
**Pull request by karolherbst**
there seems no connection between the entries we generated out of the header with the entries we find in the table when we do so.

So let split this and parse the entries right if we detect an entry based voltage table.

### CODE REVIEW

File: nvbios/nvbios.c [hunk 1]
```diff
@@ -1704,6 +1705,7 @@ int main(int argc, char **argv) {
 			entry_length = bios->data[start+2];
 			entry_count = bios->data[start+3];	// XXX: NFI what the entries are for
 			is_pwm_based = bios->data[start+4] & 1;
+			is_header_based = bios->data[start+4] & 0x2;
```
Comment (mupuf) [inline nvbios/nvbios.c]: ok, new flag, cool. What I don't like is the name because it seems to imply that it is sufficient information to know if we need to use the header or the entries, which is wrong because PWM tables will always use the header, no matter what this bit is.

So, how about gpio_header_mode or prefer_header?

How about prefer_header_src?

File: nvbios/nvbios.c [hunk 2]
```diff
@@ -1802,39 +1804,71 @@ int main(int argc, char **argv) {
...
-			} else {
+			} else if (is_header_based) {
 				step_uv = le16(start+22);
+				min_uv = le32(start+10);
+				max_uv = le32(start+14);
 				printf("-- Mode GPIO, Base voltage %d µV, voltage step %d µV, acceptable range [%d, %d] µV --\n\n",
-					volt_uv, step_uv, le32(start+10), le32(start+14));
+					volt_uv, step_uv, min_uv, max_uv);
+			} else {
+				printf("-- Mode GPIO --\n\n");
```
Comment (mupuf) [inline nvbios/nvbios.c]: Mode GPIO (entry-based)

File: nvbios/nvbios.c [hunk 3]
```diff
@@ -1802,39 +1804,71 @@ int main(int argc, char **argv) {
...
 				printf("-- Mode PWM, acceptable range [%d, %d] µV, frequency %d kHz, base voltage %d µV (unk = %d), range %d µV --\n\n",
 					le32(start+10), le32(start+14), le32(start+5) / 1000000,
 					le32(start+18) & 0xfffff,
 					le32(start+18) >> 20, le32(start+22));
-			} else {
+			} else if (is_header_based) {
 				step_uv = le16(start+22);
+				min_uv = le32(start+10);
+				max_uv = le32(start+14);
 				printf("-- Mode GPIO, Base voltage %d µV, voltage step %d µV, acceptable range [%d, %d] µV --\n\n",
```
Comment (mupuf) [inline nvbios/nvbios.c]: Mode GPIO (header-generated)

File: nvbios/nvbios.c [hunk 4]
```diff
@@ -1802,39 +1804,71 @@ int main(int argc, char **argv) {
...
-				if (is_pwm_based) {
-					printf("-- Vid %d, voltage %d µV (unk = %d) --\n", i, volt, le32(start) >> 21);
+				if (is_pwm_based || is_header_based) {
+					int v = le32(start);
+					if (v)
+						printf("-- entry %d, unk = %x --\n", i, v);
 				} else {
-					printf("-- Vid %d, voltage %d µV/%d µV (unk = %d) --\n", i, volt_uv, volt, le32(start) >> 21);
-					volt_uv += step_uv;
+					int vid = le32(start) >> 23 & 0xff;
```
Comment (mupuf) [inline nvbios/nvbios.c]: Why do a 32bits read when you want to read a 8 bit value?

File: nvbios/nvbios.c [hunk 5]
```diff
@@ -1802,39 +1804,71 @@ int main(int argc, char **argv) {
...
 					/* List the gpio tags assosiated with each voltage id */
 					int j;
 					for (j = 0; j < nr_vidtag; j++) {
-						if (!(mask & (1 << j))) {
-						/*	printf("-- Voltage unused/overridden by voltage mask --\n");*/
+						if (!(mask & (1 << j)))
 							continue;
-						}
-						printf("-- GPIO tag 0x%x(VID) data (logic %d) --\n", vidtag[j], (!!(i & (1 << j))));
+						printf("-- GPIO tag 0x%x(VID) data (logic %d) --\n", vidtag[j], (!!(vid & (1 << j))));
```
Comment (mupuf) [inline nvbios/nvbios.c]: This looks like an unrelated fix, am I right?

Please at least document it in the commit message or by adding a comment.

Comment (karolherbst) [inline nvbios/nvbios.c]: it is related. header generated means i is the index is the VID configuration on the GPIOs, but with the entry based ones we use what we read out of the entries.

File: nvbios/nvbios.c (hunk 4 above)
Comment (mupuf) [inline nvbios/nvbios.c]: I do not see any check on the version here, unlike what you proposed on Nouveau :s

Comment (karolherbst) [inline nvbios/nvbios.c]: else if (version == 0x40) {
....
} else {
    this code <== we are here
}

Comment (mupuf) [inline nvbios/nvbios.c]: sorry. Go for it then! Merge it!
